pkg/collector/tcp.go

```diff
@@ -63,48 +64,36 @@ func (cp *CollectingProcess) handleTCPClient(conn net.Conn) {
 	address := conn.RemoteAddr().String()
 	client := cp.createClient()
 	cp.addClient(address, client)
-	buff := make([]byte, cp.maxBufferSize)
 	defer cp.wg.Done()
 	defer conn.Close()
 	go func() {
+		// maxBufferSize should be good enough to hold maximum sized IPFIX packet.
+		buff := make([]byte, cp.maxBufferSize)
+		reader := bufio.NewReader(conn)
 		for {
-			size, err := conn.Read(buff)
+			length, err := getMessageLength(reader)
 			if err != nil {
-				if err == io.EOF {
-					klog.Infof("Connection from %s has been closed.", address)
-				} else {
-					klog.Errorf("Error in collecting process: %v", err)
-				}
+				klog.Errorf("error when retreiving message length: %v", err)
 				cp.deleteClient(address)
 				return
 			}
-			klog.V(2).Infof("Receiving %d bytes from %s", size, address)
-			buffBytes := make([]byte, size)
-			copy(buffBytes, buff[:size])
-			for size > 0 {
-				length, err := getMessageLength(bytes.NewBuffer(buffBytes))
-				if err != nil {
-					klog.Error(err)
-					buffBytes = nil
-					break
-				}
-				if size < length {
-					klog.Errorf("Message length %v is larger than size read from buffer %v", length, size)
-					buffBytes = nil
-					break
-				}
-				size = size - length
-				// get the message here
-				message, err := cp.decodePacket(bytes.NewBuffer(buffBytes[0:length]), address)
-				if err != nil {
-					klog.Error(err)
-					buffBytes = nil
-					break
-				}
-				klog.V(4).Infof("Processed message from exporter %v, number of records: %v, observation domain ID: %v",
-					message.GetExportAddress(), message.GetSet().GetNumberOfRecords(), message.GetObsDomainID())
-				buffBytes = buffBytes[length:]
+			_, err = io.ReadFull(reader, buff[:length])
+			if err != nil {
+				klog.Errorf("error when reading the message: %v", err)
+				cp.deleteClient(address)
+				return
+			}
+			// copiedBytes is needed to avoid race condition with the aggregation
+			// process.
+			copiedBytes := make([]byte, length)
+			copy(copiedBytes, buff[:length])
+			message, err := cp.decodePacket(bytes.NewBuffer(copiedBytes), address)
+			if err != nil {
+				klog.Error(err)
+				continue
 			}
+			klog.V(4).Infof("Processed message from exporter %v, number of records: %v, observation domain ID: %v",
+				message.GetExportAddress(), message.GetSet().GetNumberOfRecords(), message.GetObsDomainID())
```

Comment:
not in this PR, I think we should replace `message.GetExportAddress()` with other useful information since export address will be Pod host IP.

Reply:
Modified it with set type. Agree that we have to improve debugging in the intermediate process.